pkg/platform/kube/client/deployer.go

```diff
@@ -146,6 +160,278 @@ func (d *Deployer) Deploy(ctx context.Context,
+
+			secretNameLimit := 63 - (len(functionconfig.NuclioFlexVolumeSecretNamePrefix) + len(functionName) + 2)
+			if len(flexVolumeSecretName) > secretNameLimit {
+				flexVolumeSecretName = flexVolumeSecretName[:secretNameLimit]
+			}
+
+			// create the access key hash based on the reference
+			accessKeyRefHashString := functionconfig.GenerateAccessKeyRefHashString(secretValue)
+			accessKeyHashLabels = append(accessKeyHashLabels, accessKeyRefHashString)
+
+			// check if a secret with the same access key reference already exists
+			existingFlexVolumeSecrets, err := d.consumer.KubeClientSet.CoreV1().Secrets(functionNamespace).List(ctx, metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("%s=%s", common.NuclioResourceLabelKeyAccessKey, accessKeyRefHashString),
+			})
+			if err != nil {
+				return errors.Wrap(err, "Failed to list flex volume secrets")
+			}
+
+			// if a secret with the same access key reference exists, use it
+			if len(existingFlexVolumeSecrets.Items) > 0 {
+				d.logger.DebugWithCtx(ctx, "Found existing flex volume secret with same access key reference, skipping creation")
+				continue
+			}
+
+			// create the secret
+			secretConfig := &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: flexVolumeSecretName,
+					Labels: map[string]string{
+						common.NuclioResourceLabelKeyFunctionName: functionName,
+						common.NuclioResourceLabelKeyAccessKey:    accessKeyRefHashString,
+					},
+				},
+				Type: "v3io/fuse",
+				StringData: map[string]string{
+					"accessKey": secretValue,
+				},
+			}
+			if _, err := d.consumer.KubeClientSet.CoreV1().Secrets(functionNamespace).Create(ctx, secretConfig, metav1.CreateOptions{}); err != nil {
+				return errors.Wrap(err, "Failed to create flex volume secret")
+			}
+
+		}
+	}
+
+	// delete stale flex volume secrets
+	if err := d.deleteStaleFlexVolumeSecrets(ctx, accessKeyHashLabels, functionName, functionNamespace); err != nil {
+		return errors.Wrap(err, "Failed to delete stale flex volume secrets")
+	}
+
+	return nil
+}
+func (d *Deployer) deleteStaleFlexVolumeSecrets(ctx context.Context, accessKeyHashLabels []string, name, namespace string) error {
+
+	// get all secrets for the function
+	secrets, err := d.consumer.KubeClientSet.CoreV1().Secrets(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", common.NuclioResourceLabelKeyFunctionName, name),
+	})
+	if err != nil {
+		return errors.Wrap(err, "Failed to list function flex volume secrets")
+	}
+
+	// delete stale flex volume secrets
+	for _, secret := range secrets.Items {
+		if secret.Type == "v3io/fuse" {
+			if labelValue, exists := secret.Labels[common.NuclioResourceLabelKeyAccessKey]; exists &&
+				!common.StringSliceContainsString(accessKeyHashLabels, labelValue) {
+				if err := d.consumer.KubeClientSet.CoreV1().Secrets(namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{}); err != nil {
+					return errors.Wrap(err, "Failed to delete stale flex volume secret")
+				}
```

Comment:
best effort remove all secrets. if some was not deleted, aggregate its name + log a warning. eventually fail with all secrets failed to be deleted and return the aggregated err.

Reply:
I don't want to remove all secrets for a couple of reasons:
1. We handle the flex volume secret before create/update of the function secret - why delete it if we can update it?
2. I delete the stale flex volume secrets after creating new ones if needed.

While writing (2) I realize that since the secret names are unique every time, we can just delete all of the flex volume secrets and then create it again if needed.
(1) is still valid imo.